src/modules/flow/freegeoip/freegeoip.c

```diff
@@ -0,0 +1,239 @@
+    if (!endptr || *endptr)
+        return false;
+    *output = value;
+    return true;
+}
+
+static void
+freegeoip_query_finished(void *data, struct sol_http_response *response)
+{
+    struct freegeoip_data *mdata = data;
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    enum fields fields = FIELD_ALL;
+    double latitude = 0.0f;
+    double longitude = 0.0f;
+
+    RESPONSE_CHECK_API(response, mdata);
+
+    if (!response->content.used) {
+        SOL_WRN("Empty response from FreeGeoIP");
+        return;
+    }
+
+    if (response->response_code != 200) {
+        sol_flow_send_error_packet(mdata->node, EINVAL,
+            "FreeGeoIP returned an unknown response code: %d",
+            response->response_code);
+        return;
+    }
+
+#define FIELD(json_field_, flow_field_) \
+    do { \
+        if (fields & FIELD_ ## flow_field_ && sol_json_token_str_eq(&key, \
+                json_field_, strlen(json_field_))) { \
+            fields &= ~(FIELD_ ## flow_field_); \
+            sol_flow_send_string_packet(mdata->node, \
+                SOL_FLOW_NODE_TYPE_LOCATION_FREEGEOIP__OUT__ ## flow_field_, \
+                strndupa(value.start + 1, value.end - value.start - 2)); \
+            continue; \
+        } \
+    } while (0)
+
+    sol_json_scanner_init(&scanner, response->content.data, response->content.used);
+    SOL_JSON_SCANNER_OBJECT_LOOP(&scanner, &token, &key, &value, reason) {
+        FIELD("ip", IP);
+        FIELD("country_name", COUNTRY_NAME);
+        FIELD("country_code", COUNTRY_CODE);
+        FIELD("city", CITY_NAME);
+        FIELD("zip_code", ZIP_CODE);
+        FIELD("time_zone", TIMEZONE);
+
+        if (fields & FIELD_LATITUDE && sol_json_token_str_eq(&key, "latitude", strlen("latitude"))) {
+            if (json_token_to_double(&value, &latitude))
+                fields &= ~FIELD_LATITUDE;
+            continue;
+        }
+
+        if (fields & FIELD_LONGITUDE && sol_json_token_str_eq(&key, "longitude", strlen("longitude"))) {
+            if (json_token_to_double(&value, &longitude))
+                fields &= ~FIELD_LONGITUDE;
+            continue;
+        }
+    }
+
+#undef FIELD
+
+    if (reason != SOL_JSON_LOOP_REASON_OK)
+        return;
+
+    if (!(fields & (FIELD_LATITUDE | FIELD_LONGITUDE))) {
+        sol_flow_send_location_latlon_packet(mdata->node,
+            SOL_FLOW_NODE_TYPE_LOCATION_FREEGEOIP__OUT__LOCATION, latitude,
+            longitude);
+    }
+}
+
+static int
+query_addr(struct freegeoip_data *mdata, const char *addr)
+{
+    char json_endpoint[PATH_MAX];
+    int r;
+
+    r = snprintf(json_endpoint, sizeof(json_endpoint), "%s/json/%s",
+        mdata->endpoint, addr ? addr : "");
+    if (r < 0 || r >= (int)sizeof(json_endpoint)) {
+        SOL_WRN("Could not prepare endpoint");
+        return -EINVAL;
+    }
+
+    r = sol_http_client_request(SOL_HTTP_METHOD_GET, json_endpoint,
```

Comment:
uh, oh... what if the node ends before sol_http_client_request() calls back? There is no way to kill the request... we need to allow that, I'd say return a request handle here and provide `sol_http_client_request_cancel()`